mail-filter/sqlgrey: Upgrade to 1.8.0, EAPI=7, GLEP-81

Philippe23 opened the pull request:
mail-filter/sqlgrey: Upgrade to 1.8.0, EAPI=7, GLEP-81

Closes: https://bugs.gentoo.org/354869
Closes: https://github.com/gentoo/gentoo/pull/13361
Bug: https://bugs.gentoo.org/696252
Package-Manager: Portage-2.3.76, Repoman-2.3.16
Signed-off-by: Philippe Chaintreuil <gentoo_bugs_peep@parallaxshift.com>

Philippe23 commented:
Footnote: I submitted a Proxy-Maintainership request for mail-filter/slqgrey: https://bugs.gentoo.org/show_bug.cgi?id=698108

Philippe23 commented:
`repoman` created that manifest.  Not sure how it should be trimmed down aside from by hand....

Philippe23 commented:
Updated the manifest after adding ```thin-manifests = true``` to my overlay's `metadata/layout.conf`.  (Thanks for the hint on IRC juippis!)

Philippe23 commented:
Added acct-*/sqlgrey ebuilds to become GLEP-81 complaint.

blinkingtwelve commented on mail-filter/sqlgrey/sqlgrey-1.8.0.ebuild:
```diff
@@ -0,0 +1,246 @@
+	einfo "emerge --config ${PN}"
+	echo
+	ewarn "Read the documentation for more info (perldoc sqlgrey) or the"
+	ewarn "included howto ${ROOT}/usr/share/doc/${PF}/HOWTO.gz"
```
nitpick, this `.gz` is now a `.bz2`.

Philippe23 replied:
Changed it to the more general `"included HOWTO in ${ROOT}/usr/share/doc/${PF}/"` as suggested by https://devmanual.gentoo.org/ebuild-writing/common-mistakes/index.html

blinkingtwelve commented on mail-filter/sqlgrey/sqlgrey-1.8.0.ebuild:
```diff
@@ -0,0 +1,246 @@
+
+EAPI=7
+
+inherit eutils systemd user
```
and perhaps `user` can also be dropped as the account creation is now handled through `acct-user/sqlgrey` in `RDEPEND`?

Philippe23 replied:
Good catch.  Down to just `inherit systemd` now.

mgorny commented on mail-filter/sqlgrey/sqlgrey-1.8.0.ebuild:
```diff
@@ -0,0 +1,246 @@
+
+inherit systemd
+
+DESCRIPTION="SQLgrey is a postfix policy service implementing a grey-listing policy"
```
I've committed an initial EAPI 7 bump as de2c9c7cc5c8c1e28fe64de78283650e16e3b7af. Please make matching changes in your ebuild.

mgorny commented on mail-filter/sqlgrey/sqlgrey-1.8.0.ebuild:
```diff
@@ -0,0 +1,246 @@
+
+	# Check if a password is set in sqlgrey.conf
+	local SQLgrey_CONF_PWD=""
+	if [ -f "${SQLgrey_CONFIG}" ]; then
```
Always use `[[ ... ]]` form in bash.

mgorny commented on mail-filter/sqlgrey/sqlgrey-1.8.0.ebuild:
```diff
@@ -0,0 +1,246 @@
+	# Check if a password is set in sqlgrey.conf
+	local SQLgrey_CONF_PWD=""
+	if [ -f "${SQLgrey_CONFIG}" ]; then
+		if (grep -iq "^[\t ]*db_pass[\t ]*=[\t ]*.*$" ${SQLgrey_CONFIG}); then
```
The subshell serves no purpose here. Also quote path variables.

mgorny commented on mail-filter/sqlgrey/sqlgrey-1.8.0.ebuild:
```diff
@@ -0,0 +1,246 @@
+	else
+		ewarn "SQLgrey configuration missing. Exiting now."
+		echo
+		exit 0
```
Don't use `exit`. `return` instead.

mgorny commented on mail-filter/sqlgrey/sqlgrey-1.8.0.ebuild:
```diff
@@ -0,0 +1,246 @@
+	einfo "    [l] Database backend runs on localhost"
+	einfo "    [r] Database backend runs on remote host"
+	einfo "    [x] Exit"
+	echo
```
Don't mix einfo and echo.

mgorny commented on mail-filter/sqlgrey/sqlgrey-1.8.0.ebuild:
```diff
@@ -0,0 +1,246 @@
+		einfo "Generating random database user password..."
+		local SQLgrey_PWD_MATRIX="0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"
+		local SQLgrey_DB_USER_PWD=""
+		while [ "${n:=1}" -le "16" ]; do
```
Use a `for (( ... ))` loop.

mgorny commented on mail-filter/sqlgrey/sqlgrey-1.8.0.ebuild:
```diff
@@ -0,0 +1,246 @@
+		ewarn ""
+
+		einfo "Creating SQLgrey PostgreSQL database \"${SQLgrey_DB_NAME}\" and user \"${SQLgrey_DB_USER_NAME}\""
+		"${ROOT}/usr/bin/psql" -h ${SQLgrey_DB_HOST} -d template1 -U postgres -c "CREATE USER ${SQLgrey_DB_USER_NAME} WITH PASSWORD '${SQLgrey_DB_USER_PWD}' NOCREATEDB NOCREATEUSER; CREATE DATABASE ${SQLgrey_DB_NAME}; GRANT ALL PRIVILEGES ON DATABASE ${SQLgrey_DB_NAME} TO ${SQLgrey_DB_USER_NAME}; GRANT ALL PRIVILEGES ON SCHEMA public TO ${SQLgrey_DB_USER_NAME}; UPDATE pg_database SET datdba=(SELECT use ...
```
All the external commands, this one and seds below need some kind of error handling.

mgorny commented on mail-filter/sqlgrey/sqlgrey-1.8.0.ebuild:
```diff
@@ -0,0 +1,246 @@
+		"${ROOT}/usr/bin/psql" -h ${SQLgrey_DB_HOST} -d template1 -U postgres -c "CREATE USER ${SQLgrey_DB_USER_NAME} WITH PASSWORD '${SQLgrey_DB_USER_PWD}' NOCREATEDB NOCREATEUSER; CREATE DATABASE ${SQLgrey_DB_NAME}; GRANT ALL PRIVILEGES ON DATABASE ${SQLgrey_DB_NAME} TO ${SQLgrey_DB_USER_NAME}; GRANT ALL PRIVILEGES ON SCHEMA public TO ${SQLgrey_DB_USER_NAME}; UPDATE pg_database SET datdba=(SELECT use ...
+
+		einfo "Changing SQLgrey configuration in sqlgrey.conf"
+		sed -i "s:^[# ]*\(db_type[ \t]*= \).*:\1Pg:gI" ${SQLgrey_CONFIG}
```
You can combine those into one sed with multiple `-e`xpressions.

Philippe23 commented:
@mgorny Changes requested have been made, PR updated.

Deviations from requested changes:

- `PATCHES`: removed the ${P} from the 1.8.0 list because the 1.7.6 change was made upstream, (B) added/left explicit `sqlgrey-1.8.0-init-openrc.patch` because I expect future versions will want it as well, starting with 1.8.0.
- Replaced/removed most of the `echo` lines except for the two surrounding the `mysql` database creation command -- I presume these are to adjust the `mysql` command's output, not the einfo/ewarn logging.

juippis commented:
Ping @mgorny could you take another look?

Philippe23 commented:
> Ping @mgorny could you take another look?

It's been a month (granted, one filled with holidays), figured I'd ping again.  @mgorny

mgorny commented on acct-group/sqlgrey/metadata.xml:
```diff
@@ -0,0 +1,18 @@
+  </maintainer>
+
+  <maintainer type="person">
+    <email>mjo@gentoo.org</email>
```
I'm confused. Did mjo request being added here or agree on being added?

Philippe23 replied:
Good catch.  That's a copy-and-paste from another package I maintain.

Do I need an actual Gentoo dev in metadata.xml, or is it just fine with me + proxy-maint?

mgorny replied:
Just you and us are fine.

Philippe23 replied:
👍 Thanks.

mgorny commented on mail-filter/sqlgrey/sqlgrey-1.8.0.ebuild:
```diff
@@ -0,0 +1,252 @@
+
+DESCRIPTION="A postfix policy service implementing a grey-listing policy"
+HOMEPAGE="http://sqlgrey.sourceforge.net/"
+SRC_URI="mirror://sourceforge/sqlgrey/sqlgrey-1.8%20%28stable%29/${P}.tar.gz"
```
Replace that with the direct `downloads.sourceforge...` URI, this mirror no longer lists mirrors.

mgorny commented on mail-filter/sqlgrey/sqlgrey-1.8.0.ebuild:
```diff
@@ -0,0 +1,252 @@
+DEPEND="${RDEPEND}
+	sys-apps/sed"
+
+DOCS="HOWTO FAQ README README.OPTINOUT README.PERF TODO Changelog"
```
Make it an array.

mgorny commented on mail-filter/sqlgrey/sqlgrey-1.8.0.ebuild:
```diff
@@ -0,0 +1,252 @@
+
+pkg_config () {
+	# SQLgrey configuration file
+	local SQLgrey_CONFIG="${ROOT}/etc/sqlgrey/sqlgrey.conf"
```
The shell convention is to make local vars lowercase. Also `EROOT`.

mgorny commented on mail-filter/sqlgrey/sqlgrey-1.8.0.ebuild:
```diff
@@ -0,0 +1,252 @@
+	# Check if a password is set in sqlgrey.conf
+	local SQLgrey_CONF_PWD=""
+	if [[ -f "${SQLgrey_CONFIG}" ]]; then
+		if grep -iq "^[\t ]*db_pass[\t ]*=[\t ]*.*$" "${SQLgrey_CONFIG}"; then
```
Couldn't you go straight for `sed` instead of using two regexes?

mgorny commented on mail-filter/sqlgrey/sqlgrey-1.8.0.ebuild:
```diff
@@ -0,0 +1,252 @@
+
+	# Check if we need SQLgrey to configure for this system or not
+	local SQLgrey_DB_HOST="localhost"
+	local SQLgrey_KEY_INPUT="l,r,x"
```
Why use a variable that's only used once?

mgorny commented on mail-filter/sqlgrey/sqlgrey-1.8.0.ebuild:
```diff
@@ -0,0 +1,252 @@
+	local SQLgrey_DB_HOST="localhost"
+	local SQLgrey_KEY_INPUT="l,r,x"
+	einfo "SQLgrey database backend configuration"
+	einfo "  Please select where SQLgrey database will run:"
```
Why not just ask for hostname, and default it to `localhost` (via `read -i ...`).

mgorny commented on mail-filter/sqlgrey/sqlgrey-1.8.0.ebuild:
```diff
@@ -0,0 +1,252 @@
+		local SQLgrey_PWD_MATRIX="0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"
+		local SQLgrey_DB_USER_PWD=""
+		for (( n=1; n<=16; n++ )); do
+			SQLgrey_DB_USER_PWD="${SQLgrey_DB_USER_PWD}${SQLgrey_PWD_MATRIX:$(($RANDOM%${#SQLgrey_PWD_MATRIX})):1}"
```
`+=`

mgorny commented on mail-filter/sqlgrey/sqlgrey-1.8.0.ebuild:
```diff
@@ -0,0 +1,252 @@
+
+	# Generate random password
+	if [[ "${SQLgrey_CONF_PWD}" == "" ]]; then
+		einfo "Generating random database user password..."
```
Are you sure about this? Wouldn't it be better to let user choose a password?

mgorny commented on mail-filter/sqlgrey/sqlgrey-1.8.0.ebuild:
```diff
@@ -0,0 +1,252 @@
+	einfo "  Please select what kind of database you like to use:"
+	if use postgres || has_version dev-perl/DBD-Pg ; then
+		einfo "    [p] PostgreSQL"
+		SQLgrey_KEY_INPUT="${SQLgrey_KEY_INPUT},p"
```
`+=` and below.

mgorny commented on mail-filter/sqlgrey/sqlgrey-1.8.0.ebuild:
```diff
@@ -0,0 +1,252 @@
+	done
+
+	# If we don't use SQLite, the password must not be set
+	if [[ ( "${SQLgrey_DB_TYPE}" != "s" ) && ( "${SQLgrey_CONF_PWD}" != "" ) ]]; then
```
Those parentheses are unnecessary.

mgorny commented on mail-filter/sqlgrey/sqlgrey-1.8.0.ebuild:
```diff
@@ -0,0 +1,252 @@
+		ewarn ""
+
+		einfo "Creating SQLgrey PostgreSQL database \"${SQLgrey_DB_NAME}\" and user \"${SQLgrey_DB_USER_NAME}\""
+		"${ROOT}/usr/bin/psql" -h ${SQLgrey_DB_HOST} -d template1 -U postgres -c "CREATE USER ${SQLgrey_DB_USER_NAME} WITH PASSWORD '${SQLgrey_DB_USER_PWD}' NOCREATEDB NOCREATEUSER; CREATE DATABASE ${SQLgrey_DB_NAME}; GRANT ALL PRIVILEGES ON DATABASE ${SQLgrey_DB_NAME} TO ${SQLgrey_DB_USER_NAME}; GRANT ALL PRIVILEGES ON SCHEMA public TO ${SQLgrey_DB_USER_NAME}; UPDATE pg_database SET datdba=(SELECT use ...
```
EROOT, line wrapping.

mgorny commented on mail-filter/sqlgrey/sqlgrey-1.8.0.ebuild:
```diff
@@ -0,0 +1,252 @@
+		"${ROOT}/usr/bin/psql" -h ${SQLgrey_DB_HOST} -d template1 -U postgres -c "CREATE USER ${SQLgrey_DB_USER_NAME} WITH PASSWORD '${SQLgrey_DB_USER_PWD}' NOCREATEDB NOCREATEUSER; CREATE DATABASE ${SQLgrey_DB_NAME}; GRANT ALL PRIVILEGES ON DATABASE ${SQLgrey_DB_NAME} TO ${SQLgrey_DB_USER_NAME}; GRANT ALL PRIVILEGES ON SCHEMA public TO ${SQLgrey_DB_USER_NAME}; UPDATE pg_database SET datdba=(SELECT use ...
+
+		einfo "Changing SQLgrey configuration in sqlgrey.conf"
+		sed -i \
```
Can't you just put db type in a variable, and combine those `sed`s into one below ifs?

boerwastaken commented:
can I be helpful in pointing out that while indeed the previous version of the ebuild featured an elaborate (too elaborate?) `pkg_config`, for maintainability of such an ebuild it may be much better to drop such "let me write a config file for you" features and let the user take responsibility for that.
I mean, it's not like every Gentoo ebuild has a pkg_config that sets things up for you to the fullest extent possible, so I think we can change our minds as to what extent we want to do it in this ebuild.

Besides, given the expected user base of this package — mail server admins — we can entertain some hope that users will be able to copy and adapt the self-documenting example config from `/usr/share/doc/...`.

Less is more :-D

mgorny commented:
Yep, I'd also prefer that approach.

Philippe23 commented:
👍 That makes three of us.

I've removed the `pkg_config()`, made `DOCS` an array, and made the URI direct.  I also removed mjo from the acct metadata.xmls under the presumption that proxy-maint + me are sufficient.

Philippe23 commented on acct-user/sqlgrey/sqlgrey-0.ebuild:
```diff
@@ -0,0 +1,14 @@
+
+ACCT_USER_ID=336
+ACCT_USER_GROUPS=( sqlgrey )
+ACCT_USER_HOME="/var/spool/sqlgrey"
```
Let me test without and see if it causes any troubles.

I don't see anything in the default conf that references it.  Only other thought is that a home folder would be a sane location for the db if you were running SQLite.

Philippe23 replied:
I get errors at startup without a home folder, so it does appear to be needed.
